libs/aws/batch.js

```diff
@@ -115,11 +121,76 @@ export default (aws) => {
+         * Prepare to run a job by downloading the scitran snapshot and
+         * creating a mongo entry for it.
+         */
+        prepareAnalysis(job, callback) {
+            job.uploadSnapshotComplete = !!job.uploadSnapshotComplete;
+            job.analysis = {
+                analysisId: uuid.v4(),
+                status: 'UPLOADING',
+                created: new Date(),
+                attempts: 0,
+                notification: false
+            };
+            //making consistent with agave ??
+            job.appLabel = job.jobName;
+            job.appVersion = job.jobDefinition.match(/\d+$/)[0];
+
+            scitran.downloadSymlinkDataset(job.snapshotId, (err, hash) => {
+                job.datasetHash = hash;
+                job.parametersHash = crypto.createHash('md5').update(JSON.stringify(job.parameters)).digest('hex');
+
+                // jobDefintion is the full ARN including version, region, etc
+                c.crn.jobs.findOne({
+                    jobDefinition:  job.jobDefinition,
+                    datasetHash:    job.datasetHash,
+                    parametersHash: job.parametersHash,
+                    snapshotId:     job.snapshotId
+                }, {}, (err, existingJob) => {
+                    if (err) {
+                        return callback(err);
+                    }
+                    if (existingJob) {
+                        return callback({existingJob: existingJob});
+                    }
+
+                    c.crn.jobs.insertOne(job, (err, mongoJob) => {
+                        callback(err, job, mongoJob.insertedId);
+                    });
+                });
+            }, {snapshot: true});
+        },
+
+        /**
+         * Upload any required data and submit jobs to Batch API
+         *
+         * This is called by worker processes and must be process safe.
+         */
+        startAnalysis(job, jobId, userId, callback) {
+            let hash = job.datasetHash;
+            s3.uploadSnapshot(hash, () => {
+                const batchJobParams = this.buildBatchParams(job, hash);
+
+                this.startBatchJobs(batchJobParams, jobId, (err) => {
+                    if (err) {
+                        // This is an unexpected error, probably from batch.
+                        console.log(err);
+                        // Cleanup the failed to submit job
+                        // TODO - Maybe we save the error message into another field for display?
+                        c.crn.jobs.updateOne({_id: jobId}, {$set: {'analysis.status': 'REJECTED'}});
+                    } else {
+                        emitter.emit(events.JOB_STARTED, {job: batchJobParams, createdDate: job.analysis.created}, userId);
+                    }
+                    callback();
+                });
+            });
+        },
+
+        /**
+         * Start AWS Batch Jobs for an analysis
          */
-        startBatchJob(batchJob, jobId, callback) {
```

Comment:
This breaks retry now because aws.batch.startBatchJob(batchJobParams..) is still being called on retry.